DigitalEcosystem/data_splitting/bg_extra_descriptors.py

```diff
@@ -0,0 +1,327 @@
+
+
+# In[]:
+
+
+desc = GlobalInstabilityIndex()
+
+def maybe_global_instability(struct):
+    try:
+        return desc.featurize(struct)[0]
+    except:
+        return None
+
+with warnings.catch_warnings():
+    warnings.simplefilter('ignore')
+    df['global_instability'] = df.ox_struct.swifter.apply(maybe_global_instability).copy()
+
+
+# In[]:
+
+
+neighbor_finder = JmolNN()
+
+def average_bond_length(structure, featurizer = AverageBondLength(neighbor_finder)):
+    n_atoms = len(structure)
+    try:
+        lengths = map(lambda i: featurizer.featurize(structure, i)[0], range(n_atoms))
+        return sum(lengths) / n_atoms
+    except IndexError:
+        return None
+
+df['bond_length_average'] = df.ox_struct.swifter.apply(average_bond_length).copy()
+
+
+# In[]:
+
+
+from matminer.featurizers.site.bonding import AverageBondAngle
+
+def average_bond_angle(structure, featurizer = AverageBondAngle(neighbor_finder)):
+    n_atoms = len(structure)
+    try:
+        angles = map(lambda i: featurizer.featurize(structure, i)[0], range(n_atoms))
+        return sum(angles) / n_atoms
+    except IndexError:
+        return None
+df['bond_angle_average'] = df.ox_struct.swifter.apply(average_bond_angle).copy()
+
+
+# In[]:
+
+
+def average_cn(structure, neighbor_finder = neighbor_finder):
+    n_atoms = len(structure)
+    cns = map(lambda i: neighbor_finder.get_cn(structure, i), range(n_atoms))
+    return sum(cns) / n_atoms
+df['average_cn'] = df.ox_struct.swifter.apply(average_cn).copy()
+
+
+# In[]:
+
+
+def ab_perimeter_area_ratio(structure):
+    a, b, c = structure.lattice.matrix
+    perimeter = 2*np.linalg.norm(a) + 2*np.linalg.norm(b)
+    area = np.linalg.norm(np.cross(a,b))
+    return perimeter / area
+df['perimeter_area_ratio'] = df.ox_struct.swifter.apply(ab_perimeter_area_ratio).copy()
+
+
+# In[]:
+
+
+df["formula"] = df["atoms_object (unitless)"].swifter.apply(lambda atoms: atoms.get_chemical_formula(empirical=True))
+
+
+# In[]:
+
+
+df.to_pickle("all_descriptors.pkl")
+
+object_cols = ["atoms_object (unitless)",
+               "ox_struct"]
+
+regression_irrelevant = object_cols + [
+    'discovery_process (unitless)',
+    'potcars (unitless)',
+    'is_hubbard (unitless)',
+    'energy_per_atom (eV)',
+    'decomposition_energy (eV/atom)',
+    'is_bandgap_direct (unitless)',
+    'is_metal (unitless)',
+    'energy_vdw_per_atom (eV/atom)',
+    'total_magnetization (Bohr Magneton)']
+
+df.drop(columns=object_cols).to_csv("both_extra_descriptors.csv")
+
+
+# # Chemical Info
+
+# In[]:
+
+
+for allowed_symbols, set_name in [
+    [["N"], "nitrides"],
+    [["O"], "oxides"],
+    [["S"], "sulfides"],
+    [["C"], "carbides"],
+    [["F", "Cl", "Br", "Cl"], "halides"],
+    [["O", "S", "Se", "Te"], "chalcogens"],
+    [["N", "O", "As", "Sb"], "pnictogens"],
+    [["Al", "Ga", "Ge", "In", "Sn", "Sb", "Tl", "Pb", "Bi", "Po"], "pBlockMetals"]
```

Comment:
The naming convention as above can be defined and reused from a separate file too.